### PR TITLE
Added map fields for Python

### PR DESCRIPTION
**Pull request by haberman**

### CODE REVIEW

File: python/map.c [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+  Py_INCREF(parent);
+  return &map->ob_base;
+}
+
+void PyUpb_MapContainer_Reify(PyObject* _self, upb_map* map) {
+  PyUpb_MapContainer* self = (PyUpb_MapContainer*)_self;
+  PyUpb_ObjCache_Add(map, &self->ob_base);
+  Py_DECREF(self->ptr.parent);
+  self->ptr.map = map;  // Overwrites self->ptr.parent.
+  self->field = self->field & ~(uintptr_t)1;
```
Comment (fowles) [inline python/map.c:106]: `self->field &= ~(uintptr_t)1`

File: python/map.c [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+  upb_arena* arena = PyUpb_Arena_Get(self->arena);
+  upb_msgval u_key, u_val;
+  if (!PyUpb_PyToUpb(key, key_f, &u_key, arena)) return -1;
+
+  if (val) {
+    if (!PyUpb_PyToUpb(val, val_f, &u_val, arena)) return -1;
+    upb_map_set(map, u_key, u_val, arena);
+  } else {
+    if (!upb_map_delete(map, u_key)) {
+      PyErr_Format(PyExc_KeyError, "Key not present in map");
```
Comment (fowles) [inline python/map.c:149]: do the other branches not need to set an error?

File: python/map.c [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+  } else {
+    Py_RETURN_FALSE;
+  }
+}
+
+PyObject* PyUpb_MapContainer_Clear(PyObject* _self, PyObject* key) {
+  PyUpb_MapContainer* self = (PyUpb_MapContainer*)_self;
+  upb_map* map = PyUpb_MapContainer_GetIfWritable(self);
+  if (map) upb_map_clear(map);
+  Py_RETURN_NONE;
```
Comment (fowles) [inline python/map.c:176]: shouldn't this do something for the non-reified case?  If feels like you should still clear something there...

File: python/map.c [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+                                        PyObject* kwargs) {
+  PyUpb_MapContainer* self = (PyUpb_MapContainer*)_self;
+  static const char* kwlist[] = {"key", "default", NULL};
+  PyObject* key;
+  PyObject* default_value = NULL;
+  upb_map* map = PyUpb_MapContainer_GetIfWritable(self);
+  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "O|O", (char**)kwlist, &key,
+                                   &default_value)) {
+    return NULL;
+  }
```
Comment (fowles) [inline python/map.c:189]: after this point, don't `key` and `default_value` require decrefs?

File: python/map.c [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+  if (map && upb_map_get(map, u_key, &u_val)) {
+    return PyUpb_UpbToPy(u_val, val_f, self->arena);
+  } else {
+    if (default_value) {
+      Py_INCREF(default_value);
+      return default_value;
+    } else {
+      Py_RETURN_NONE;
+    }
+  }
```
Comment (fowles) [inline python/map.c:207]: can you unnest this control flow?
```
if (!PyUpb...) return NULL;

if (map && map_get) return ...

if (default_value) {
  inc
 retrun default_value;
}

Py_RETURN_NONE;
```

File: python/map.c [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+  PyUpb_MapContainer* self = (PyUpb_MapContainer*)_self;
+  const upb_fielddef* f = PyUpb_MapContainer_GetField(self);
+  const upb_msgdef* entry_m = upb_fielddef_msgsubdef(f);
+  return PyUpb_Descriptor_GetClass(entry_m);
+}
+
+Py_ssize_t PyUpb_MapContainer_Length(PyObject* _self) {
+  PyUpb_MapContainer* self = (PyUpb_MapContainer*)_self;
+  upb_map* map = PyUpb_MapContainer_GetIfWritable(self);
+  return map ? upb_map_size(map) : 0;
```
Comment (fowles) [inline python/map.c:221]: feels weird that the non-reified length is reported as 0

File: python/map.c [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+
+  Py_RETURN_NONE;
+}
+
+static PyObject* PyUpb_MapContainer_Repr(PyObject* _self) {
+  PyUpb_MapContainer* self = (PyUpb_MapContainer*)_self;
+  upb_map* map = PyUpb_MapContainer_GetIfWritable(self);
+  PyObject* dict = PyDict_New();
+  if (map) {
+    size_t iter = UPB_MAP_BEGIN;
```
Comment (fowles) [inline python/map.c:257]: can you move iter to be cuddled up against the loop?  I like the practice of keeping the iteration pattern very close together for easier grep and scan

File: python/map.c [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+    const upb_fielddef* f = PyUpb_MapContainer_GetField(self);
+    const upb_msgdef* entry_m = upb_fielddef_msgsubdef(f);
+    const upb_fielddef* key_f = upb_msgdef_field(entry_m, 0);
+    const upb_fielddef* val_f = upb_msgdef_field(entry_m, 1);
+    while (upb_mapiter_next(map, &iter)) {
+      PyObject* key =
+          PyUpb_UpbToPy(upb_mapiter_key(map, iter), key_f, self->arena);
+      PyObject* val =
+          PyUpb_UpbToPy(upb_mapiter_value(map, iter), val_f, self->arena);
```
Comment (fowles) [inline python/map.c:266]: can either of these fail?

File: python/map.c [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+      Py_DECREF(key);
+      Py_DECREF(val);
+    }
+  }
+  PyObject* repr = PyObject_Repr(dict);
+  Py_DECREF(dict);
+  return repr;
+}
+
+PyObject* PyUpb_MapContainer_Subscript(PyObject* _self, PyObject* key) {
```
Comment (fowles) [inline python/map.c:277]: can you move AssignSubscript and Subscript to be next to each other in the source?

File: python/map.c [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+  }
+  return PyUpb_UpbToPy(u_val, val_f, self->arena);
+}
+
+PyObject* PyUpb_MapContainer_GetOrCreateWrapper(upb_map* map,
+                                                const upb_fielddef* f,
+                                                PyObject* arena) {
+  PyObject* ret = PyUpb_ObjCache_Get(map);
+
+  if (!ret) {
```
Comment (fowles) [inline python/map.c:305]: invert this conditional and early return

File: python/protobuf.c [hunk 11]
```diff
@@ -254,8 +255,16 @@ static const char *PyUpb_GetClassName(PyType_Spec *spec) {
...
+  int result = PyModule_AddObject(m, name, type);
+  return type && result == 0 ? (PyTypeObject*)type : NULL;
+}
+
+PyTypeObject* PyUpb_AddClassWithBases(PyObject* m, PyType_Spec* spec,
+                                      PyObject* bases) {
+  PyObject* type = PyType_FromSpecWithBases(spec, bases);
+  const char* name = PyUpb_GetClassName(spec);
+  int result = PyModule_AddObject(m, name, type);
+  return type && result == 0 ? (PyTypeObject*)type : NULL;
```
Comment (fowles) [inline python/protobuf.c:267]: I would rather push this cast into the variable declaration and have a simple ternary

File: upb/msg.c [hunk 12]
```diff
@@ -285,6 +285,7 @@ static int _upb_mapsorter_cmpstr(const void *_a, const void *_b) {
   return a.size - b.size;
 }
 
+#include <stdio.h>
```
Comment (fowles) [inline upb/msg.c:288]: this feels weird

File: python/map.c (hunk 1 above)
Comment (haberman) [reply python/map.c:106]: Done.

File: python/map.c (hunk 2 above)
Comment (haberman) [reply python/map.c:149]: No, `PyUpb_PyToUpb()` will set the error if it fails.

File: python/map.c (hunk 3 above)
Comment (haberman) [reply python/map.c:176]: I think you are right. Like repeated fields, I think it should `AssureWritable()` first, to reify and clear.

File: python/map.c (hunk 4 above)
Comment (haberman) [reply python/map.c:189]: No, the local variables have "borrowed references", meaning that the refcount was not increased.

File: python/map.c (hunk 5 above)
Comment (haberman) [reply python/map.c:207]: Done.

File: python/map.c (hunk 6 above)
Comment (haberman) [reply python/map.c:221]: Why is that weird? A non-reified object behaves like an empty object until mutated.

File: python/map.c (hunk 7 above)
Comment (haberman) [reply python/map.c:257]: Done.

File: python/map.c (hunk 8 above)
Comment (haberman) [reply python/map.c:266]: Done.

File: python/map.c (hunk 9 above)
Comment (haberman) [reply python/map.c:277]: I had been alphabetizing as the organizational principle here, but I can see the benefit of breaking that convention for this case.

File: python/map.c (hunk 10 above)
Comment (haberman) [reply python/map.c:305]: Done.

File: python/protobuf.c (hunk 11 above)
Comment (haberman) [reply python/protobuf.c:267]: If the declaration is `PyObject` then other uses of `type` require a cast. I realized I needed to rework the error handling on this a bit, take a look at the new code and see what you think.

File: upb/msg.c (hunk 12 above)
Comment (haberman) [reply upb/msg.c:288]: Residual printf debugging artifact, removed.

**Comment by haberman** [comment]
PTAL.

File: python/map.c [hunk 13]
```diff
@@ -254,16 +275,21 @@ static PyObject* PyUpb_MapContainer_Repr(PyObject* _self) {
...
+    size_t iter = UPB_MAP_BEGIN;
     while (upb_mapiter_next(map, &iter)) {
       PyObject* key =
           PyUpb_UpbToPy(upb_mapiter_key(map, iter), key_f, self->arena);
       PyObject* val =
           PyUpb_UpbToPy(upb_mapiter_value(map, iter), val_f, self->arena);
+      if (!key || !val) {
+        Py_XDECREF(key);
+        Py_XDECREF(val);
+        return NULL;
```
Comment (fowles) [inline python/map.c:291]: I think you need to `PY_DECREF(dict)` as well

File: python/protobuf.c [hunk 14]
```diff
@@ -253,18 +253,24 @@ static const char *PyUpb_GetClassName(PyType_Spec *spec) {
 }
 
 PyTypeObject *PyUpb_AddClass(PyObject *m, PyType_Spec *spec) {
-  PyObject *type = PyType_FromSpec(spec);
+  PyObject *type = (void*)PyType_FromSpec(spec);
```
Comment (fowles) [inline python/protobuf.c:256]: you added the `(void*)` here but didn't change the lhs of the object.  I am guessing you don't need this cast.

File: python/map.c (hunk 6 above)
Comment (fowles) [reply python/map.c:242]: feels a bit like only clearing when it is not reified.  Why should "ask length, try to get the thing, ask length again" return different values for the length?

File: python/map.c (hunk 13 above)
Comment (haberman) [reply python/map.c:291]: Done.

File: python/protobuf.c (hunk 14 above)
Comment (haberman) [reply python/protobuf.c:256]: Done.

**Comment by haberman** [comment]
PTAL.

**Review by fowles — APPROVED** [review]

**Pull request merged**